File: ET3ALEM_SERVER/BusinessEntities/Models/Quiz.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace BusinessEntities.Models
{
    public class Quiz : IEquatable<Quiz>
    {
        public Quiz()
        {
        }

        public int Id { get; set; }
        public string Code { get; set; }

        [Required] public string Name { get; set; }

        public string Instructions { get; set; }
        public virtual List<QuizQuestion> QuizQuestions { get; set; } = new();
        public virtual List<QuizAttempt> QuizAttempts { get; set; }
        public int TotalGrade { get; set; }
        public DateTime? StartDate { get; set; }
        public DateTime? EndDate { get; set; }
        public DateTime CreatedDate { get; set; }
        public bool NoDueDate { get; set; }
        public int DurationSeconds { get; set; }
        public bool UnlimitedTime { get; set; }
        [Required] public string UserId { get; set; }
        public virtual User User { get; set; }
        public int AllowedAttempts { get; set; }
        public bool UnlimitedAttempts { get; set; }
        public bool ShowGrade { get; set; }
        public bool AutoGrade { get; set; }
        public bool ShuffleQuestions { get; set; }
        public bool IncludeAllQuestions { get; set; }
        public int? IncludedQuestionsCount { get; set; }

        public bool Equals(Quiz other)
        {
            if (ReferenceEquals(null, other)) return false;
            if (ReferenceEquals(this, other)) return true;
            return Id == other.Id && string.Equals(Code, other.Code, StringComparison.InvariantCultureIgnoreCase);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((Quiz) obj);
        }

        public override int GetHashCode()
        {
            var hashCode = new HashCode();
            hashCode.Add(Id);
            hashCode.Add(Code, StringComparer.InvariantCultureIgnoreCase);
            return hashCode.ToHashCode();
        }

        public static bool operator ==(Quiz left, Quiz right)
        {
            return Equals(left, right);
        }

        public static bool operator !=(Quiz left, Quiz right)
        {
            return !Equals(left, right);
        }
    }
}

File: ET3ALEM_CLIENT_WEB/src/app/quiz/Model/quiz.ts
import { Expose, Type } from 'class-transformer';
import { QuizQuestion } from './quizQuestion';

export class Quiz {
    Id: number;
    readonly UserId: string = null;
    Name: string;
    Code: string;
    Instructions: string;
    DurationSeconds: number;
    UnlimitedTime: boolean;
    @Type(() => Date)
    StartDate: Date;
    @Type(() => Date)
    EndDate: Date;
    @Type(() => Date)
    CreatedDate: Date;
    NoDueDate: boolean;
    AllowedAttempts: number;
    UnlimitedAttempts: boolean;
    ShowGrade: boolean;
    AutoGrade: boolean;
    TotalGrade: number;
    ShuffleQuestions: boolean;
    IncludeAllQuestions: boolean;
    IncludedQuestionsCount: number;
    @Type(() => QuizQuestion)
    QuizQuestions: Array<QuizQuestion> = [];

    constructor(
        Id: number = 0,
        code: string = '',
        title: string,
        instructions: string = '',
        durationSeconds: number,
        UnlimitedTime: boolean,
        dateStart: Date,
        dueEnd: Date,
        createdDate: Date,
        noDueDate: boolean,
        QuizQuestions?: Array<QuizQuestion>,
        AllowedAttempts: number = 1,
        UnlimitedAttempts: boolean = false,
        ShowGrade: boolean = true,
        AutoGrade: boolean = true,
        ShuffleQuestions: boolean = false,
        IncludeAllQuestions: boolean = true,
        IncludedQuestionsCount: number = null) {
        this.Id = Id;
        this.Code = code;
        this.Name = title;
        this.Instructions = instructions;
        this.DurationSeconds = durationSeconds;
        this.UnlimitedTime = UnlimitedTime;
        this.StartDate = dateStart;
        this.EndDate = dueEnd;
        this.NoDueDate = noDueDate;
        this.CreatedDate = createdDate;
        this.QuizQuestions = QuizQuestions;
        if (!this.UserId)
            this.UserId = 'NA';
        this.AllowedAttempts = AllowedAttempts;
        this.UnlimitedAttempts = UnlimitedAttempts;
        this.ShowGrade = ShowGrade,
            this.AutoGrade = AutoGrade;
        this.ShuffleQuestions = ShuffleQuestions;
        this.IncludeAllQuestions = IncludeAllQuestions || IncludedQuestionsCount == this.QuizQuestions?.length;
        this.IncludedQuestionsCount = IncludedQuestionsCount;
    }

    public updateQuiz(title: string,
        instructions: string = '',
        durationSeconds: number,
        UnlimitedTime: boolean,
        dateStart: Date,
        dueEnd: Date,
        noDueDate: boolean,
        QuizQuestions?: Array<QuizQuestion>,
        AllowedAttempts: number = 1,
        UnlimitedAttempts: boolean = false,
        ShowGrade: boolean = true,
        AutoGrade: boolean = true,
        ShuffleQuestions: boolean = false,
        IncludeAllQuestions: boolean = true,
        IncludedQuestionsCount: number = null) {
        this.Name = title;
        this.Instructions = instructions;
        this.DurationSeconds = durationSeconds;
        this.UnlimitedTime = UnlimitedTime;
        this.StartDate = dateStart;
        this.EndDate = dueEnd;
        this.NoDueDate = noDueDate;
        this.QuizQuestions = QuizQuestions;
        this.AllowedAttempts = AllowedAttempts;
        this.UnlimitedAttempts = UnlimitedAttempts;
        this.ShowGrade = ShowGrade,
            this.AutoGrade = AutoGrade;
        this.ShuffleQuestions = ShuffleQuestions;
        this.IncludeAllQuestions = IncludeAllQuestions || IncludedQuestionsCount == this.QuizQuestions?.length;;
    }
}